Pre requisites

a-zakem opened the pull request:
Fix #100 #37 

Adding PreRequisite checks.  Check if user has modified/deleted starter code or tried to import external libraries, and give appropriate feedback.

eyurko commented on client/services/code_evaluators/PreRequisiteCheckDispatcherService.js:
```diff
@@ -0,0 +1,34 @@
+      }
+    };
+  }
+]);
```
Add newline

eyurko replied:
Add newline

eyurko commented on client/domain/PreRequisiteEvalResultObjectFactory.js:
```diff
@@ -0,0 +1,52 @@
+    };
+
+    // Instance methods
+    PreRequisiteEvalResult.prototype.getCode = function() {
```
If you're also using the CodeEvalResult, do you think you need to store the code here, too? Or maybe store the code on the snapshot and just have the eval results contain error messages? I'm just spitballing, here.

seanlip replied:
Hm, I think we can do that in a separate pass if needed -- I've added it to my radar. Thanks for noting this!

seanlip replied:
Hi @a-zakem,

I just realized something -- the code that's used for PrerequisiteEvalResult is not actually the same as the code in the CodeEvalResult, right? If I'm not mistaken, the first one is the raw code and the second is the code after a bunch of extra testing stuff is added to it...

a-zakem replied:
Hi Sean,

I think that depends on when the CodeEvalResult object is created.  If the
CodeEvalResult object is created as a result of a syntax error during the
the initial run of the raw code, then the CodeEvalResult object will
contain the raw code.  If the student submission passes the syntax error
check and proceeds to pre-processing, then the CodeEvalResult object will
contain the pre-processed code including the testing code.

In the PreRequisites branch as merged, PreRequisiteEvalResultObjectFactory
was renamed CodePrereqCheckResultObjectFactory, and the
CodePrereqCheckResult does not store the submitted code (raw or
otherwise).  At present, I believe the stored code in the CodeEvalResult is
only used in the FeedbackGeneratorService's _getBuggyOutputTestFeedback
function, to check that the student has submitted different code before
incrementing the hintIndex for the same error.

seanlip replied:
Ah ok -- I checked SolutionHandlerService, and the codeEvalResult is created after running the preprocessed code, so it would be different from the code stored in the prereq check result object (if we had indeed stored code in it). Since we didn't, though, that's a moot point, and I think what is currently in master looks good. We can always add the raw code later to the prereq result if we need it.

Given this, I don't think we need to pull the code out of CodeEvalResult, either now or in the future. Would you agree?

a-zakem replied:
For my two cents, I think it is fine to leave the code in CodeEvalResult at present, as there isn't a need to access the student code that caused a PrereqCheckFailure at some later point in time.  If there is a feature in the future that could use that information, the submitted code could be pulled out of CodeEvalResult and added to Snapshot instead.

seanlip replied:
OK, sounds good -- though the code in PrereqCheckFailure would still be different from the code in CodeEvalResult, so we might want to still keep the code with the results of the run.

In any case, though, this depends on what we do in the future, so let's punt the decision. Thanks for your thoughts!

eyurko commented on client/services/FeedbackGeneratorService.js:
```diff
@@ -221,6 +221,23 @@ tie.factory('FeedbackGeneratorService', [
+        var errorMessage =
+          preReqEvalResult.getPreReqErrorMessage();
+        feedback.appendTextParagraph(
+          "Oops!  There is a problem with your code:");
```
I wouldn't say "Oops!"; it's a bit too informal. Go for "It looks like there's a problem with your code: "

eyurko commented on client/services/SolutionHandlerService.js:
```diff
@@ -19,53 +19,74 @@
-              potentialSyntaxErrorString);
+          task, starterCode, studentCode, auxiliaryCode, language) {
+        // First, check pre-requisites for the submitted code
+        return PreRequisiteCheckDispatcherService.checkCode(
```
As a nit, I think it should be just Prerequisite, not PreRequisite

eyurko commented on client/services/code_evaluators/PythonPreRequisiteCheckService.js:
```diff
@@ -0,0 +1,123 @@
+          continue;
+        }
+        //handle "return" statements in starter code
+        if (starterCodeLine.replace(/^\s+/,"").startsWith("return ")) {
```
Always pull out regexes as constants at the start of the file. Makes it more obvious to the person reading your code.

eyurko commented on client/services/code_evaluators/PythonPreRequisiteCheckService.js:
```diff
@@ -0,0 +1,123 @@
+          var errorMessage = ['It looks like you deleted or modified ',
+            'the starter code!  Our evaluation program requires the ',
+            'function names given in the starter code.  Here\'s the '
+            ,'starter code again:'].join('');
```
I think we also have a reset button, now? Might be worth mentioning, if that's a thing that exists.

eyurko commented on client/services/code_evaluators/PythonPreRequisiteCheckServiceSpec.js:
```diff
@@ -0,0 +1,102 @@
+			expect(codeLibs).toEqual([]);
+		});
+	});
+});
```
Add newline at EOF

seanlip commented on client/domain/PrerequisiteEvalResultObjectFactory.js:
```diff
@@ -0,0 +1,52 @@
+    var PrerequisiteEvalResult = function(
+      code, prereqErrorMessage, starterCode) {
+      this._code = code;
+      this._prereqErrorMessage = prereqErrorMessage;
```
This is a bit confusing. The purpose of this domain object would presumably be to store enough information to reconstruct the result of the prerequisite check.

So, presumably the only field here should be a list called something like this._prerequisiteCheckFailures. Each entry is a dict with a 'type' key (whose value is PREREQ_CHECK_TYPE_BAD_IMPORT = 'badImport' or PREREQ_CHECK_TYPE_MISSING_STARTER_CODE = 'missingStarterCode' -- these named constants can be declared in app.js). If the key is PREREQ_CHECK_TYPE_MISSING_STARTER_CODE then the dict also has an extra 'expectedStarterCode' key; you may need something similar for badImport as well.

The error messages themselves should presumably go in the service which generates the feedback objects based on the essential info in this domain object.

seanlip commented on client/services/SolutionHandlerService.js:
```diff
@@ -19,53 +19,74 @@
-  'FeedbackGeneratorService', 'SnapshotObjectFactory', 'TranscriptService',
-  'CodeSubmissionObjectFactory',
+  'FeedbackGeneratorService', 'PrerequisiteCheckDispatcherService',
+  'SnapshotObjectFactory', 'TranscriptService', 'CodeSubmissionObjectFactory',
```
Nit: let's move 'CodeSubmissionObjectFactory' to the line below for parity with line 27. Easier to check that the dependencies line up, that way.

seanlip commented on client/services/code_evaluators/PythonPrerequisiteCheckService.js:
```diff
@@ -0,0 +1,126 @@
+
+    const LEFT_TRIM_PATTERN = /^\s+/;
+
+    var checkStarterCodePresent = function(starterCode, code) {
```
This is far too complicated / ad hoc. It'll be hard to maintain.

So, just to verify: the point of this prereq check is to ensure that the functions in the starter code still exist within the learner's code, right? If so, could we instead restructure the logic in this function roughly as follows?

```
    expectedTopLevelFunctions = extractTopLevelFunctions(starterCode);
    verifyFunctionsExist(code, expectedTopLevelFunctions);
```

seanlip commented on client/services/code_evaluators/PythonPrerequisiteCheckService.js:
```diff
@@ -0,0 +1,126 @@
+      return true;
+    };
+
+    var getCodeLibs = function(code) {
```
getImportedLibraries

seanlip commented on client/services/code_evaluators/PythonPrerequisiteCheckService.js:
```diff
@@ -0,0 +1,126 @@
+
+    var getCodeLibs = function(code) {
+      var codeLines = code.split('\n');
+      var coddLibs = [];
```
Spelling is wrong. Might want to use importedLibraries, anyway.

seanlip commented on client/services/code_evaluators/PythonPrerequisiteCheckService.js:
```diff
@@ -0,0 +1,126 @@
+      var codeLines = code.split('\n');
+      var coddLibs = [];
+      var pattern = new RegExp('^import\\ \\w+$');
+      for(var i=0; i<codeLines.length; i++) {
```
I'm surprised that the linter isn't catching this, but there should be single spaces around = and '<', as well as before '('. Ditto elsewhere.

seanlip commented on client/services/code_evaluators/PythonPrerequisiteCheckService.js:
```diff
@@ -0,0 +1,126 @@
+      for(var i=0; i<codeLines.length; i++) {
+        if(pattern.test(codeLines[i])) {
+          var words = codeLines[i].split(' ');
+          coddLibs.push(words[1]);
```
You should be able to get the pattern directly from the regexp check by using capture groups. I recommend doing that instead, it's much cleaner.

seanlip commented on client/services/code_evaluators/PythonPrerequisiteCheckService.js:
```diff
@@ -0,0 +1,126 @@
+      // Returns a promise.
+      checkCode: function(starterCode, code) {
+        if (!(checkStarterCodePresent(starterCode, code))) {
+          var errorMessage = ['It looks like you deleted or modified ',
```
This stuff should go in the service that creates the feedback messages that get displayed to the user. The object storing the prereq check result should only contain the error type and the minimal amount of info needed to reproduce what check actually failed.

seanlip commented on client/services/code_evaluators/PythonPrerequisiteCheckService.js:
```diff
@@ -0,0 +1,126 @@
+                        'The following libraries are not ',
+                        'supported:\n'].join('');
+        codeLibs = getCodeLibs(code);
+        var extLibs = [];
```
Again, all this stuff is way too complicated. You could instead do something like this:

    var unsupportedImports = importedLibraries.filter(function(importedLibrary) {
      return PYTHON_STANDARD_LIBRARIES.indexOf(importedLibrary) === -1;
    });

Make sure that the importedLibraries list is already trimmed, etc. Then pass the unsupportedImports as a customization to the prereq check result factory.

seanlip commented on client/services/code_evaluators/PythonPrerequisiteCheckServiceSpec.js:
```diff
@@ -0,0 +1,102 @@
+ */
+
+describe('PythonPrerequisiteCheckService', function() {
+	var PythonPrerequisiteCheckService;
```
Use spaces (and 2-space indentation), not tabs. Again I'm surprised the linter isn't catching this, I'll take a look. Is this synced with the latest version of master?

seanlip commented on client/services/code_evaluators/PythonPrerequisiteCheckServiceSpec.js:
```diff
@@ -0,0 +1,102 @@
+				PythonPrerequisiteCheckService.getCodeLibs(code);
+			expect(codeLibs).toEqual(['numpy','pandas']);
+		});
+		it('does not capture strings with \'import\' in text', function() {
```
Nice test!

seanlip commented on client/services/code_evaluators/PythonPrerequisiteCheckServiceSpec.js:
```diff
@@ -0,0 +1,102 @@
+						starterCode, code);
+				expect(starterCodePresent).toEqual(true);
+		});
+		it('returns false when starter code lines are present, but out of order',
```
Eh, I'm not sure the order matters...

a-zakem commented:
Hi Sean,

I can confirm that the linter is not catching the style errors you pointed
out; I fetched from origin and merged in the master branch, but was still
able to push while receiving a "Linter tests passed" message.

I'll work on fixing the issues above.

--Aaron

On Tue, Apr 25, 2017 at 4:25 PM, Sean Lip <notifications@github.com> wrote:

> *@seanlip* requested changes on this pull request.
>
> Thanks! I've taken a full pass -- a lot of the parts are there, but
> there's some improvements that can be made to make things conceptually
> clear. PTAL and let me know if you have questions on the comments I've
> written.
> ------------------------------
>
> In client/app.js
> <https://github.com/google/tie/pull/237#discussion_r113329092>:
>
> > @@ -33,6 +33,9 @@ tie.constant('CODE_EXECUTION_TIMEOUT_SECONDS', 3);
>  // Supported languages.
>  tie.constant('LANGUAGE_PYTHON', 'python');
>  tie.constant('ALL_SUPPORTED_LANGUAGES', ['python']);
> +tie.constant('PYTHON_STANDARD_LIBRARIES', ['collections', 'document', 'image',
>
> I'm not sure we need document, processing, or turtle.
> ------------------------------
>
> In client/domain/SnapshotObjectFactory.js
> <https://github.com/google/tie/pull/237#discussion_r113329510>:
>
> >        this._codeEvalResult = codeEvalResult;
> +      this._prereqEvalResult = prereqEvalResult;
>
> For ease of understanding, let's put codePrereqCheckResult (see other
> comment above re naming) above codeEvalResult, since it comes first
> chronologically.
> ------------------------------
>
> In client/domain/PrerequisiteEvalResultObjectFactory.js
> <https://github.com/google/tie/pull/237#discussion_r113330157>:
>
> > +/**
> + * @fileoverview Factory for creating new frontend instances of
> + * PreRequisiteCheckResult domain objects.
> + */
> +
> +tie.factory('PrerequisiteEvalResultObjectFactory', [
> +  function() {
> +    var PrerequisiteEvalResult = function(
> +      code, prereqErrorMessage, starterCode) {
> +      this._code = code;
> +      this._prereqErrorMessage = prereqErrorMessage;
> +      this._starterCode = starterCode;
> +    };
> +
> +    // Instance methods
> +    PrerequisiteEvalResult.prototype.getCode = function() {
>
> Where is this used? Why do we need to store the code at all?
> ------------------------------
>
> In client/domain/PrerequisiteEvalResultObjectFactory.js
> <https://github.com/google/tie/pull/237#discussion_r113331076>:
>
> > +// You may obtain a copy of the License at
> +//
> +//      http://www.apache.org/licenses/LICENSE-2.0
> +//
> +// Unless required by applicable law or agreed to in writing, software
> +// distributed under the License is distributed on an "AS-IS" BASIS,
> +// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> +// See the License for the specific language governing permissions and
> +// limitations under the License.
> +
> +/**
> + * @fileoverview Factory for creating new frontend instances of
> + * PreRequisiteCheckResult domain objects.
> + */
> +
> +tie.factory('PrerequisiteEvalResultObjectFactory', [
>
> How about calling this CodePrereqCheckResult, for parity with
> CodeEvalResult? The difficulty with PrerequisiteEvalResult is that, in
> CodeEvalResult, it's the code that's being evaluated, which implies that
> here the prereqs are being evaluated.
> ------------------------------
>
> In client/domain/PrerequisiteEvalResultObjectFactory.js
> <https://github.com/google/tie/pull/237#discussion_r113331382>:
>
> > +//
> +// Licensed under the Apache License, Version 2.0 (the "License");
> +// you may not use this file except in compliance with the License.
> +// You may obtain a copy of the License at
> +//
> +//      http://www.apache.org/licenses/LICENSE-2.0
> +//
> +// Unless required by applicable law or agreed to in writing, software
> +// distributed under the License is distributed on an "AS-IS" BASIS,
> +// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> +// See the License for the specific language governing permissions and
> +// limitations under the License.
> +
> +/**
> + * @fileoverview Factory for creating new frontend instances of
> + * PreRequisiteCheckResult domain objects.
>
> The capitalization that's being used here does not match the actual name
> of the domain object.
> ------------------------------
>
> In client/domain/PrerequisiteEvalResultObjectFactory.js
> <https://github.com/google/tie/pull/237#discussion_r113332110>:
>
> > +// distributed under the License is distributed on an "AS-IS" BASIS,
> +// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> +// See the License for the specific language governing permissions and
> +// limitations under the License.
> +
> +/**
> + * @fileoverview Factory for creating new frontend instances of
> + * PreRequisiteCheckResult domain objects.
> + */
> +
> +tie.factory('PrerequisiteEvalResultObjectFactory', [
> +  function() {
> +    var PrerequisiteEvalResult = function(
> +      code, prereqErrorMessage, starterCode) {
> +      this._code = code;
> +      this._prereqErrorMessage = prereqErrorMessage;
>
> This is a bit confusing. The purpose of this domain object would
> presumably be to store enough information to reconstruct the result of the
> prerequisite check.
>
> So, presumably the only field here should be a list called something like
> this._prerequisiteCheckFailures. Each entry is a dict with a 'type' key
> (whose value is PREREQ_CHECK_TYPE_BAD_IMPORT = 'badImport' or
> PREREQ_CHECK_TYPE_MISSING_STARTER_CODE = 'missingStarterCode' -- these
> named constants can be declared in app.js). If the key is
> PREREQ_CHECK_TYPE_MISSING_STARTER_CODE then the dict also has an extra
> 'expectedStarterCode' key; you may need something similar for badImport as
> well.
>
> The error messages themselves should presumably go in the service which
> generates the feedback objects based on the essential info in this domain
> object.
> ------------------------------
>
> In client/services/SolutionHandlerService.js
> <https://github.com/google/tie/pull/237#discussion_r113332260>:
>
> > @@ -19,53 +19,74 @@
>
>  tie.factory('SolutionHandlerService', [
>    '$q', 'CodePreprocessorDispatcherService', 'CodeRunnerDispatcherService',
> -  'FeedbackGeneratorService', 'SnapshotObjectFactory', 'TranscriptService',
> -  'CodeSubmissionObjectFactory',
> +  'FeedbackGeneratorService', 'PrerequisiteCheckDispatcherService',
> +  'SnapshotObjectFactory', 'TranscriptService', 'CodeSubmissionObjectFactory',
>
> Nit: let's move 'CodeSubmissionObjectFactory' to the line below for parity
> with line 27. Easier to check that the dependencies line up, that way.
> ------------------------------
>
> In client/services/code_evaluators/PythonPrerequisiteCheckService.js
> <https://github.com/google/tie/pull/237#discussion_r113333270>:
>
> > +// limitations under the License.
> +
> +/**
> + * @fileoverview Service for performing pre-requisite checks on Python code snippets.
> + */
> +
> +tie.factory('PythonPrerequisiteCheckService', [
> +  'PYTHON_STANDARD_LIBRARIES',
> +  'PrerequisiteEvalResultObjectFactory',
> +  function(
> +      PYTHON_STANDARD_LIBRARIES,
> +      PrerequisiteEvalResultObjectFactory) {
> +
> +    const LEFT_TRIM_PATTERN = /^\s+/;
> +
> +    var checkStarterCodePresent = function(starterCode, code) {
>
> This is far too complicated / ad hoc. It'll be hard to maintain.
>
> So, just to verify: the point of this prereq check is to ensure that the
> functions in the starter code still exist within the learner's code, right?
> If so, could we instead restructure the logic in this function roughly as
> follows?
>
>     expectedTopLevelFunctions = extractTopLevelFunctions(starterCode);
>     verifyFunctionsExist(code, expectedTopLevelFunctions);
>
> ------------------------------
>
> In client/services/code_evaluators/PythonPrerequisiteCheckService.js
> <https://github.com/google/tie/pull/237#discussion_r113333430>:
>
> > +        else {
> +          while (codeLineIndex < codeLines.length) {
> +            if (codeLines[codeLineIndex] === starterCodeLine) {
> +              break;
> +            }
> +            codeLineIndex++;
> +          }
> +        }
> +        if (codeLineIndex >= codeLines.length && i < starterCodeLines.length) {
> +          return false;
> +        }
> +      }
> +      return true;
> +    };
> +
> +    var getCodeLibs = function(code) {
>
> getImportedLibraries
> ------------------------------
>
> In client/services/code_evaluators/PythonPrerequisiteCheckService.js
> <https://github.com/google/tie/pull/237#discussion_r113333478>:
>
> > +            if (codeLines[codeLineIndex] === starterCodeLine) {
> +              break;
> +            }
> +            codeLineIndex++;
> +          }
> +        }
> +        if (codeLineIndex >= codeLines.length && i < starterCodeLines.length) {
> +          return false;
> +        }
> +      }
> +      return true;
> +    };
> +
> +    var getCodeLibs = function(code) {
> +      var codeLines = code.split('\n');
> +      var coddLibs = [];
>
> Spelling is wrong. Might want to use importedLibraries, anyway.
> ------------------------------
>
> In client/services/code_evaluators/PythonPrerequisiteCheckService.js
> <https://github.com/google/tie/pull/237#discussion_r113333538>:
>
> > +            }
> +            codeLineIndex++;
> +          }
> +        }
> +        if (codeLineIndex >= codeLines.length && i < starterCodeLines.length) {
> +          return false;
> +        }
> +      }
> +      return true;
> +    };
> +
> +    var getCodeLibs = function(code) {
> +      var codeLines = code.split('\n');
> +      var coddLibs = [];
> +      var pattern = new RegExp('^import\\ \\w+$');
> +      for(var i=0; i<codeLines.length; i++) {
>
> I'm surprised that the linter isn't catching this, but there should be
> single spaces around = and '<', as well as before '('. Ditto elsewhere.
> ------------------------------
>
> In client/services/code_evaluators/PythonPrerequisiteCheckService.js
> <https://github.com/google/tie/pull/237#discussion_r113333702>:
>
> > +        }
> +        if (codeLineIndex >= codeLines.length && i < starterCodeLines.length) {
> +          return false;
> +        }
> +      }
> +      return true;
> +    };
> +
> +    var getCodeLibs = function(code) {
> +      var codeLines = code.split('\n');
> +      var coddLibs = [];
> +      var pattern = new RegExp('^import\\ \\w+$');
> +      for(var i=0; i<codeLines.length; i++) {
> +        if(pattern.test(codeLines[i])) {
> +          var words = codeLines[i].split(' ');
> +          coddLibs.push(words[1]);
>
> You should be able to get the pattern directly from the regexp check by
> using capture groups. I recommend doing that instead, it's much cleaner.
> ------------------------------
>
> In client/services/code_evaluators/PythonPrerequisiteCheckService.js
> <https://github.com/google/tie/pull/237#discussion_r113333867>:
>
> > +      var coddLibs = [];
> +      var pattern = new RegExp('^import\\ \\w+$');
> +      for(var i=0; i<codeLines.length; i++) {
> +        if(pattern.test(codeLines[i])) {
> +          var words = codeLines[i].split(' ');
> +          coddLibs.push(words[1]);
> +        }
> +      }
> +      return coddLibs;
> +    };
> +
> +    return {
> +      // Returns a promise.
> +      checkCode: function(starterCode, code) {
> +        if (!(checkStarterCodePresent(starterCode, code))) {
> +          var errorMessage = ['It looks like you deleted or modified ',
>
> This stuff should go in the service that creates the feedback messages
> that get displayed to the user. The object storing the prereq check result
> should only contain the error type and the minimal amount of info needed to
> reproduce what check actually failed.
> ------------------------------
>
> In client/services/code_evaluators/PythonPrerequisiteCheckService.js
> <https://github.com/google/tie/pull/237#discussion_r113334833>:
>
> > +          var errorMessage = ['It looks like you deleted or modified ',
> +            'the starter code!  Our evaluation program requires the ',
> +            'function names given in the starter code.  You can press the ',
> +            '\'Reset Code\' button below to start over.  Or, you can copy ',
> +            'the starter code below:'].join('');
> +          return Promise.resolve(
> +            PrerequisiteEvalResultObjectFactory.create(
> +              code, errorMessage, starterCode));
> +        }
> +
> +        var libErrorMessage = ['It looks like you are importing an external ',
> +                        'library.  Only standard libraries are supported.  ',
> +                        'The following libraries are not ',
> +                        'supported:\n'].join('');
> +        codeLibs = getCodeLibs(code);
> +        var extLibs = [];
>
> Again, all this stuff is way too complicated. You could instead do
> something like this:
>
> var unsupportedImports = importedLibraries.filter(function(importedLibrary) {
>   return PYTHON_STANDARD_LIBRARIES.indexOf(importedLibrary) === -1;
> });
>
> Make sure that the importedLibraries list is already trimmed, etc. Then
> pass the unsupportedImports as a customization to the prereq check result
> factory.
> ------------------------------
>
> In client/services/code_evaluators/PythonPrerequisiteCheckServiceSpec.js
> <https://github.com/google/tie/pull/237#discussion_r113334996>:
>
> > +// You may obtain a copy of the License at
> +//
> +//      http://www.apache.org/licenses/LICENSE-2.0
> +//
> +// Unless required by applicable law or agreed to in writing, software
> +// distributed under the License is distributed on an "AS-IS" BASIS,
> +// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> +// See the License for the specific language governing permissions and
> +// limitations under the License.
> +
> +/**
> + * @fileoverview Unit tests for the PythonPreRequisiteCheckService.
> + */
> +
> +describe('PythonPrerequisiteCheckService', function() {
> +	var PythonPrerequisiteCheckService;
>
> Use spaces (and 2-space indentation), not tabs. Again I'm surprised the
> linter isn't catching this, I'll take a look. Is this synced with the
> latest version of master?
> ------------------------------
>
> In client/services/code_evaluators/PythonPrerequisiteCheckServiceSpec.js
> <https://github.com/google/tie/pull/237#discussion_r113335091>:
>
> > +					PythonPrerequisiteCheckService.checkStarterCodePresent(
> +						starterCode, code);
> +				expect(starterCodePresent).toEqual(false);
> +		});
> +	});
> +	describe('getCodeLibs', function() {
> +		it('correctly parses import statements', function() {
> +			var code = ['import numpy\n',
> +				'import pandas\n', '\n',
> +				'def myFunction(arg): \n',
> +				'\treturn arg'].join('');
> +			var codeLibs =
> +				PythonPrerequisiteCheckService.getCodeLibs(code);
> +			expect(codeLibs).toEqual(['numpy','pandas']);
> +		});
> +		it('does not capture strings with \'import\' in text', function() {
>
> Nice test!
> ------------------------------
>
> In client/services/code_evaluators/PythonPrerequisiteCheckServiceSpec.js
> <https://github.com/google/tie/pull/237#discussion_r113335132>:
>
> > +				var starterCodePresent =
> +					PythonPrerequisiteCheckService.checkStarterCodePresent(
> +						starterCode, code);
> +				expect(starterCodePresent).toEqual(false);
> +		});
> +		it('returns true when only argument to \'return\' is modified)',
> +			function() {
> +				var code = ['def myFunction(arg):\n',
> +					'\tresult = arg.rstrip()\n',
> +					'\treturn myResult\n'].join('');
> +				var starterCodePresent =
> +					PythonPrerequisiteCheckService.checkStarterCodePresent(
> +						starterCode, code);
> +				expect(starterCodePresent).toEqual(true);
> +		});
> +		it('returns false when starter code lines are present, but out of order',
>
> Eh, I'm not sure the order matters...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/tie/pull/237#pullrequestreview-34707776>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AaKvlFRta_Zqkg60EGUrkEDKrIFSJzyUks5rzoDQgaJpZM4NCfoR>
> .
>

seanlip commented:
Thanks, @a-zakem! OK, good to know -- I'll investigate the linter issues.

seanlip commented on client/domain/SnapshotObjectFactory.js:
```diff
@@ -19,13 +19,22 @@
 
     // Instance methods.
+    Snapshot.prototype.getCodePrereqCheckResult = function() {
+      return this._prereqEvalResult;
```
I think you forgot to update this.

a-zakem replied:
Fixed.

seanlip commented on client/domain/SnapshotObjectFactory.js:
```diff
@@ -19,13 +19,22 @@
+    };
+
+    Snapshot.prototype.setCodePrereqCheckResult = function(prereqEvalResult) {
+      this._prereqEvalResult = prereqEvalResult;
```
Ditto. The names are wrong.

a-zakem replied:
Fixed.

seanlip commented on client/services/FeedbackGeneratorService.js:
```diff
@@ -222,6 +225,43 @@ tie.factory('FeedbackGeneratorService', [
       },
+      getPrereqFailureFeedback: function(codePrereqCheckResult) {
+        var feedback = FeedbackObjectFactory.create(false);
+        var preReqFailures = codePrereqCheckResult.getPrereqCheckFailures();
```
Here and elsewhere: be careful of casing. "Prerequisite" is one word, and so this should be `prereqFailures` (or, for parity with the method name, `prereqCheckFailures`).

a-zakem replied:
Fixed.

seanlip commented on client/services/FeedbackGeneratorService.js:
```diff
@@ -222,6 +225,43 @@ tie.factory('FeedbackGeneratorService', [
+        var feedback = FeedbackObjectFactory.create(false);
+        var preReqFailures = codePrereqCheckResult.getPrereqCheckFailures();
+
+        /* This function shouldn't be invoked unless there is at least one
```
I don't understand. Wouldn't it make more sense to throw an Error in this case if we require the input arg to have at least one element?

a-zakem replied:
Changed to throw an Error.

seanlip commented on client/services/FeedbackGeneratorService.js:
```diff
@@ -222,6 +225,43 @@ tie.factory('FeedbackGeneratorService', [
+
+        // Check first error type and generate appropriate feedback message.
+        var preReqFailure = preReqFailures[0];
+        if (preReqFailure.get('type') === PREREQ_CHECK_TYPE_MISSING_STARTER_CODE) {
```
There ought to be a domain object for prereqFailure, and we should be able to call prereqFailure.isMissingStarterCode(), prereqFailure.isBadImport().

a-zakem replied:
Added PrereqCheckFailureObjectFactory under client/domain with functions PrereqCheckFailure.isMissingStarterCode() and PrereqCheckFailure.isBadImport().

seanlip commented on client/services/FeedbackGeneratorService.js:
```diff
@@ -222,6 +225,43 @@ tie.factory('FeedbackGeneratorService', [
+          feedback.appendCodeParagraph(badImports.join('\n'));
+          return feedback;
+        } else {
+          /* If pre-requisite check type is not handled above, generate generic
```
No -- throw an Error.

a-zakem replied:
Changed to throw an Error.

seanlip commented on client/services/FeedbackGeneratorService.js:
```diff
@@ -222,6 +225,43 @@ tie.factory('FeedbackGeneratorService', [
+            'function names given in the starter code.  You can press the ',
+            '\'Reset Code\' button below to start over.  Or, you can copy ',
+            'the starter code below:'].join(''));
+          feedback.appendCodeParagraph(preReqFailure.get('starterCode'));
```
Should be prereqFailure.getStarterCode(), and similarly for badImports.

a-zakem replied:
Added functions prereqCheckFailure.getStarterCode() and prereqCheckFailure.getBadImports() to PrereqCheckFailure object.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,119 @@
+      for (var i = 0; i < starterCodeLines.length; i++) {
+        var line = starterCodeLines[i].trim();
+        if (line.startsWith("def ")) {
+          topLevelFunctions.push(line);
```
Let's call this topLevelFunctionLines, otherwise it looks like it's storing just the function names.

a-zakem replied:
Changed.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,119 @@
+    };
+
+    var verifyTopLevelFunctionsExist = function(
+      code, expectedTopLevelFunctions) {
```
expectedTopLevelFunctionLines (similarly)

a-zakem replied:
Changed.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,119 @@
+      for (var i = 0; i < codeLines.length; i++) {
+        codeLines[i] = codeLines[i].trim();
+      }
+      for (var i = 0; i < expectedTopLevelFunctions.length; i++) {;
```
```
return expectedTopLevelFunctionLines.every(function(expectedLine) {
  return codeLines.indexOf(expectedLine) !== -1;
});

a-zakem replied:
Changed.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,119 @@
+      code, expectedTopLevelFunctions) {
+      var codeLines = code.split('\n');
+      for (var i = 0; i < codeLines.length; i++) {
+        codeLines[i] = codeLines[i].trim();
```
Why trim, btw? I thought we'd want no whitespace at the start, otherwise we risk matching inner functions.

a-zakem replied:
I mistakenly thought Javascript's trim() was equivalent to Python's rstrip().  I added a function rightTrim() to PythonPrereqCheckService which uses regular expressions and string replace to trim white space at the end of a string.    I'm using rightTrim() so that additional white space at the end of the line in either the user's code or starter code will not trigger a prereq failure for missing starter code function lines (although it should probably trigger a style warning in the future).

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,119 @@
+      return topLevelFunctions;
+    };
+
+    var verifyTopLevelFunctionsExist = function(
```
Nit: let's call this doTopLevelFunctionLinesExist()

(I'd tend to expect verify() to throw an exception if the check fails. The proposed new name makes it clearer that it returns a boolean.)

a-zakem replied:
Changed.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,119 @@
+    var getImportedLibraries = function(code) {
+      var codeLines = code.split('\n');
+      var importedLibraries = [];
+      var importPattern = new RegExp('(^import\\ )(\\w+)($)');
```
You don't really need some of these capture groups. I.e. you could do:

    new RegExp('^import\\ (\\w+)$');

a-zakem replied:
Changed.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,119 @@
+    var getUnsupportedImports = function(importedLibraries) {
+      var unsupportedImports = importedLibraries.filter(function(
+        importedLibrary) {
+        return PYTHON_STANDARD_LIBRARIES.indexOf(importedLibrary) === -1;
```
Indent this line by 2.

Also, I forgot to mention this earlier, but can we rename the constant to SUPPORTED_PYTHON_LIBRARIES or SUPPORTED_PYTHON_LIBS? PYTHON_STANDARD_LIBRARIES kind of implies "all standard libraries in python".

a-zakem replied:
Changed.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,119 @@
+      checkCode: function(starterCode, code) {
+        // check that starter code is present
+        if (!(checkStarterCodePresent(starterCode, code))) {
+          var prereqCheckFailures = [{'type':
```
There should be a domain object for the inner dicts -- don't use raw JS dicts to pass around the info.

a-zakem replied:
Added PrereqCheckFailure domain object.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckServiceSpec.js:
```diff
@@ -0,0 +1,119 @@
+ */
+
+describe('PythonPrereqCheckService', function() {
+	var PythonPrereqCheckService;
```
As mentioned previously, please use spaces, not tabs. You can double-check how it looks at https://github.com/google/tie/pull/237/files#diff-8011e749cdc5745088f463a875d42ccdR19

a-zakem replied:
Changed.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckServiceSpec.js:
```diff
@@ -0,0 +1,119 @@
+					PythonPrereqCheckService.checkStarterCodePresent(
+						starterCode, code);
+				expect(starterCodePresent).toEqual(false);
+		});
```
I mentioned before that the test "it returns false when starter code lines are present, but out of order" seems odd, but I should have been clearer. I think this is a good test to include here, but we should instead check that it returns `true`.

a-zakem replied:
Added test.

seanlip commented:
Thank you, @a-zakem! On the whole, this is looking way better. I have another round of comments.

Also, one request: please could you reply to all comments? In particular, for comments that you've addressed, please reply "Done." -- this helps keep track of which comments have been addressed and which haven't. (I think you missed one.)

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,126 @@
+      SUPPORTED_PYTHON_LIBS) {
+
+    var rightTrim = function(str) {
+      //remove trailing white space at end of string
```
Nit: format as
```
// Remove trailing white space at end of string.
```
(start with capital letter, have full stop at end, have space after "//")

a-zakem replied:
Fixed.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,126 @@
+      var topLevelFunctionLines = [];
+      for (var i = 0; i < starterCodeLines.length; i++) {
+        var line = rightTrim(starterCodeLines[i]);
+        if (line.startsWith("def ")) {
```
Nit: use single quotes for consistency with rest of codebase

a-zakem replied:
Fixed.

seanlip commented on client/services/code_evaluators/PythonPrereqCheckService.js:
```diff
@@ -0,0 +1,126 @@
+    return {
+      // Returns a promise.
+      checkCode: function(starterCode, code) {
+        // check that starter code is present
```
Same comments as above for the styling of all these comments (start with capital letter, end with period).

a-zakem replied:
Fixed.

seanlip commented:
@eyurko -- can you PTAL and approve if you have no further comments? Then I'll handle the merge.

Thanks!

a-zakem commented:
Hi Sean,

I think that depends on when the CodeEvalResult object is created.  If the
CodeEvalResult object is created as a result of a syntax error during the
the initial run of the raw code, then the CodeEvalResult object will
contain the raw code.  If the student submission passes the syntax error
check and proceeds to pre-processing, then the CodeEvalResult object will
contain the pre-processed code including the testing code.

In the PreRequisites branch as merged, PreRequisiteEvalResultObjectFactory
was renamed CodePrereqCheckResultObjectFactory, and the
CodePrereqCheckResult does not store the submitted code (raw or
otherwise).  At present, I believe the stored code in the CodeEvalResult is
only used in the FeedbackGeneratorService's _getBuggyOutputTestFeedback
function, to check that the student has submitted different code before
incrementing the hintIndex for the same error.

On Thu, Apr 27, 2017 at 4:47 PM, Sean Lip <notifications@github.com> wrote:

> *@seanlip* commented on this pull request.
> ------------------------------
>
> In client/domain/PreRequisiteEvalResultObjectFactory.js
> <https://github.com/google/tie/pull/237#discussion_r113826884>:
>
> > +/**
> + * @fileoverview Factory for creating new frontend instances of
> + * PreRequisiteCheckResult domain objects.
> + */
> +
> +tie.factory('PreRequisiteEvalResultObjectFactory', [
> +  function() {
> +    var PreRequisiteEvalResult = function(
> +      code, preReqErrorMessage, starterCode) {
> +      this._code = code;
> +      this._preReqErrorMessage = preReqErrorMessage;
> +      this._starterCode = starterCode;
> +    };
> +
> +    // Instance methods
> +    PreRequisiteEvalResult.prototype.getCode = function() {
>
> Hi @a-zakem <https://github.com/a-zakem>,
>
> I just realized something -- the code that's used for
> PrerequisiteEvalResult is not actually the same as the code in the
> CodeEvalResult, right? If I'm not mistaken, the first one is the raw code
> and the second is the code after a bunch of extra testing stuff is added to
> it...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/tie/pull/237#discussion_r113826884>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AaKvlJyQzqG0E8W0kRndnROoQ57FbkLhks5r0Sj1gaJpZM4NCfoR>
> .
>
